python-checks/src/main/java/org/sonar/python/checks/SklearnCachedPipelineDontAccessTransformersCheck.java

```diff
@@ -0,0 +1,204 @@
+      .map(e -> ((ListLiteral) e).elements().expressions())
+      .stream()
+      .flatMap(Collection::stream)
+      .filter(e -> e.is(Tree.Kind.TUPLE));
+
+    return tuples
+      .map(t -> ((TupleImpl) t).elements())
+      .filter(e -> e.size() == 2)
+      .filter(e -> e.get(1).is(Tree.Kind.NAME))
+      .map(e2 -> {
+        if (e2.get(0).is(Tree.Kind.STRING_LITERAL)) {
+          nameToStepName.put((Name) e2.get(1), ((StringLiteral) e2.get(0)).trimmedQuotesValue());
+        }
+        return e2;
+      })
+      .map(e -> e.get(1))
+      .map(Name.class::cast);
+  }
+
+  private static Optional<PythonQuickFix> getQuickFix(Name pipelineBindingVariable, Tree name, QualifiedExpression qualifiedExpression, Map<Name, String> nameToStepName) {
+    var quickFix = PythonQuickFix.newQuickFix("Access the property through the ");
+    String stepName = nameToStepName.get(name);
+    if (stepName == null) {
+      return Optional.empty();
+    }
+    quickFix.addTextEdit(TextEditUtils.replace(qualifiedExpression.qualifier(), String.format("%s.named_steps[\"%s\"]", pipelineBindingVariable.name(), stepName)));
+    return Optional.of(quickFix.build());
+  }
+
+  private static Map<Tree, QualifiedExpression> symbolIsUsedInQualifiedExpression(Name name) {
+    Symbol symbol = name.symbol();
+    if (symbol == null) {
+      return new HashMap<>();
+    }
+    Map<Tree, QualifiedExpression> qualifiedExpressions = new HashMap<>();
+    symbol.usages().stream()
+      .filter(u -> u.tree().parent().is(Tree.Kind.QUALIFIED_EXPR))
+      .forEach(u -> qualifiedExpressions.put(((QualifiedExpression) u.tree().parent()).qualifier(), ((QualifiedExpression) u.tree().parent())));
+
+    return qualifiedExpressions;
+  }
+
+  private enum PipelineCreation {
+    PIPELINE,
+    MAKE_PIPELINE,
+    NOT_PIPELINE
+  }
+
+  private static PipelineCreation isPipelineCreation(CallExpression callExpression) {
+    return Optional.ofNullable(callExpression.calleeSymbol()).map(Symbol::fullyQualifiedName)
+      .map(fqn -> {
+        if ("sklearn.pipeline.Pipeline".equals(fqn)) {
+          return PipelineCreation.PIPELINE;
+        }
+        if ("sklearn.pipeline.make_pipeline".equals(fqn)) {
+          return PipelineCreation.MAKE_PIPELINE;
+        }
+        return PipelineCreation.NOT_PIPELINE;
+      }).orElse(PipelineCreation.NOT_PIPELINE);
+  }
+
+  private static Optional<Name> getAssignedName(Expression expression) {
+    if (expression.is(Tree.Kind.NAME)) {
+      return Optional.of((Name) expression);
+    }
+    if (expression.is(Tree.Kind.QUALIFIED_EXPR)) {
+      return getAssignedName(((QualifiedExpression) expression).name());
+    }
+    var assignment = (AssignmentStatement) TreeUtils.firstAncestorOfKind(expression, Tree.Kind.ASSIGNMENT_STMT);
+    if (assignment == null) {
+      return Optional.empty();
+    }
+    var expressions = SymbolUtils.assignmentsLhs(assignment);
+    if (expressions.size() != 1) {
```

Comment:
could it be possible for the expressions to have no elements?

Reply:
I have thought about it, and I don't know how to make one